twins/mongodb/twins.go

```diff
@@ -0,0 +1,229 @@
+	"context"
+
+	"github.com/mainflux/mainflux/twins"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+)
+
+const (
+	maxNameSize            = 1024
+	twinsCollection string = "twins"
+)
+
+type twinRepository struct {
+	db *mongo.Database
+}
+
+var _ twins.TwinRepository = (*twinRepository)(nil)
+
+// NewTwinRepository instantiates a MongoDB implementation of twin
+// repository.
+func NewTwinRepository(db *mongo.Database) twins.TwinRepository {
+	return &twinRepository{
+		db: db,
+	}
+}
+
+// Save persists the twin
+func (tr *twinRepository) Save(ctx context.Context, tw twins.Twin) (string, error) {
+	if len(tw.Name) > maxNameSize {
+		return "", twins.ErrMalformedEntity
+	}
+
+	coll := tr.db.Collection(twinsCollection)
+
+	if _, err := tr.RetrieveByID(ctx, tw.ID); err == nil {
+		return "", twins.ErrConflict
+	}
+	if _, err := tr.RetrieveByKey(ctx, tw.Key); err == nil {
+		return "", twins.ErrConflict
+	}
+
+	if _, err := coll.InsertOne(context.Background(), tw); err != nil {
+		return "", err
+	}
+
+	return tw.ID, nil
+}
+
+// Update performs an update to the existing twins. A non-nil error is
+// returned to indicate operation failure.
+func (tr *twinRepository) Update(ctx context.Context, tw twins.Twin) error {
+	if len(tw.Name) > maxNameSize {
+		return twins.ErrMalformedEntity
+	}
+
+	coll := tr.db.Collection(twinsCollection)
+
+	filter := bson.D{{"id", tw.ID}}
+	update := bson.D{{"$set", tw}}
+	res, err := coll.UpdateOne(context.Background(), filter, update)
+	if err != nil {
+		return err
+	}
+
+	if res.ModifiedCount < 1 {
+		return twins.ErrNotFound
+	}
+
+	return nil
+}
+
+// RetrieveByID retrieves the twin having the provided identifier
+func (tr *twinRepository) RetrieveByID(_ context.Context, id string) (twins.Twin, error) {
+	coll := tr.db.Collection(twinsCollection)
+	var tw twins.Twin
+
+	filter := bson.D{{"id", id}}
+	if err := coll.FindOne(context.Background(), filter).Decode(&tw); err != nil {
+		return tw, twins.ErrNotFound
+	}
+
+	return tw, nil
+}
+
+// RetrieveByKey retrieves the twin having the provided key
+func (tr *twinRepository) RetrieveByKey(_ context.Context, key string) (string, error) {
+	coll := tr.db.Collection(twinsCollection)
+	var tw twins.Twin
+
+	filter := bson.D{{"key", key}}
+	if err := coll.FindOne(context.Background(), filter).Decode(&tw); err != nil {
+		return "", twins.ErrNotFound
+	}
+
+	return tw.ID, nil
+}
+
+func (tr *twinRepository) RetrieveByThing(ctx context.Context, thingid string) (twins.Twin, error) {
+	coll := tr.db.Collection(twinsCollection)
+	tw := twins.Twin{}
+	filter := bson.D{{"thingid", thingid}}
+	if err := coll.FindOne(context.Background(), filter).Decode(&tw); err != nil {
+		return tw, twins.ErrNotFound
+	}
+
+	return tw, nil
+}
+
+func decodeTwins(ctx context.Context, cur *mongo.Cursor) ([]twins.Twin, error) {
+	defer cur.Close(ctx)
+	var results []twins.Twin
+	for cur.Next(ctx) {
+		var elem twins.Twin
```

Comment:
Create elem var before loop, do not allocate new var for every pass through the loop.

Reply:
We need a separate instance of Twin, because we append it to an array of twins. Can't use the same structure instance for every instance of twin.